torchao/quantization/quant_api.py

```diff
@@ -2382,27 +2389,140 @@ def _fpx_weight_only_transform(
+
+    "_default" is ignored for parameter replacement.
+
+    Note:
+        - The order of patterns in the OrderedDict matters as the first matching pattern is applied
+        - Parameters that are already TorchAOBaseTensor instances are skipped to avoid double quantization
     """
 
-    module_fqn_to_config: Dict[str, Optional[AOBaseConfig]] = field(
-        default_factory=dict
+    module_or_param_fqn_to_config: OrderedDictType[str, Optional[AOBaseConfig]] = field(
+        default_factory=OrderedDict
     )
 
     def __post_init__(self):
-        torch._C._log_api_usage_once("torchao.quantization.ModuleFqnToConfig")
+        torch._C._log_api_usage_once("torchao.quantization.ModuleOrParamFqnToConfig")
+
+    @property
+    def module_fqn_to_config(self):
+        """Compatibility property to maintain interface consistency with ModuleFqnToConfig."""
+        return self.module_or_param_fqn_to_config
+
+
+# maintain BC
+ModuleFqnToConfig = ModuleOrParamFqnToConfig
+
+
+def _param_fqn_to_config_handler(
+    mod_containing_param: torch.nn.Module, fqn: str, config: ModuleOrParamFqnToConfig
+):
+    """This function processes parameters within a module and applies quantization configurations
+    when the parameter's fully qualified name matches patterns defined in the config.
+
+    Args:
+        mod_containing_param (torch.nn.Module): The module containing parameters to be processed.
+        fqn (str): The fully qualified name of the module containing the parameters.
+        config (ModuleOrParamFqnToConfig): Configuration object containing regex patterns mapped
+            to quantization configurations.
+
+    Returns:
+        torch.nn.Module: The modified module with quantized parameters.
+
+    Note:
+        - Only processes top-level parameters (those directly accessible as module attributes)
+        - Skips parameters that are already TorchAOBaseTensor instances to avoid double quantization
+        - Uses the first matching pattern for each parameter
+        - Sets quantized parameters as non-differentiable (requires_grad=False)
+
+    Raises:
+        NotImplementedError: If a configuration type doesn't have a registered parameter handler.
+    """
+    top_level_named_parameters_list = [
+        (name, param)
+        for name, param in mod_containing_param.named_parameters()
+        if name in dir(mod_containing_param)
+    ]
+
+    # return if modified previously by module flow
+    if any(
+        isinstance(param, TorchAOBaseTensor)
+        for _, param in top_level_named_parameters_list
+    ):
+        return mod_containing_param
+
+    for name, param in top_level_named_parameters_list:
+        for pattern, param_config in config.module_or_param_fqn_to_config.items():
+            full_param_fqn = f"{fqn}.{name}"
+            if (pattern == full_param_fqn) or (
+                pattern[:3] == "re:" and re.search(pattern[3:], f"{fqn}.{name}")
```

Comment:
should we use `re.fullmatch`? since that's the behavior we want right?

Reply:
I also updated https://github.com/pytorch/ao/pull/3084/files we can align the implementation to check regex as well